satpy/composites/crefl_utils.py

```diff
@@ -331,6 +363,74 @@ def get_atm_variables(mus, muv, phi, height, coeffs):
     return sphalb, rhoray, TtotraytH2O, tOG
 
 
+def get_atm_variables_abi(mus, muv, phi, height, coeffs, G_O3, G_H2O, G_O2):
+    import dask
+    # coeffs = (ah2o, aO2, ao3, tau)
+    (ah2o, ao2, ao3, tau) = coeffs
+    TAUSTEP4SPHALB = .0003
+    xfd = 0.958725775
+    xbeta2 = 0.5
+    as0 = [0.33243832, 0.16285370, -0.30924818, -0.10324388, 0.11493334,
+           -6.777104e-02, 1.577425e-03, -1.240906e-02, 3.241678e-02, -3.503695e-02]
+    as1 = [0.19666292, -5.439061e-02]
+    as2 = [0.14545937, -2.910845e-02]
+    tau_step = da.linspace(TAUSTEP4SPHALB, MAXNUMSPHALBVALUES * TAUSTEP4SPHALB,
+                           MAXNUMSPHALBVALUES, chunks=int(MAXNUMSPHALBVALUES/2))
+    sphalb0 = csalbr(tau_step)
+    # phios = phi + 180.0 DONE IN PREPROCESSOR
+    xcos1 = 1.0
+    xcos2 = da.cos(xu.deg2rad(phi))
+    xcos3 = da.cos(xu.deg2rad(2.0 * phi))
+    xph1 = 1.0 + (3.0 * mus * mus - 1.0) * (3.0 * muv * muv - 1.0) * xfd / 8.0
+    xph2 = -xfd * xbeta2 * 1.5 * mus * muv * da.sqrt(1.0 - mus * mus) * da.sqrt(1.0 - muv * muv)
+    xph3 = xfd * xbeta2 * 0.375 * (1.0 - mus * mus) * (1.0 - muv * muv)
+
+    fs01 = (as0[0] + (mus + muv)*as0[1] + (mus * muv)*as0[2] + (mus * mus + muv * muv)*as0[3]
+            + (mus * mus * muv * muv)*as0[4])
+    fs02 = (as0[5] + (mus + muv)*as0[6] + (mus * muv)*as0[7] + (mus * mus + muv * muv)*as0[8]
+            + (mus * mus * muv * muv)*as0[9])
+
+    LOG.debug("Processing band:")
+    taur = tau * da.exp(-height / SCALEHEIGHT)
+    xlntaur = da.log(taur)
+    fs0 = fs01 + fs02 * xlntaur
+    fs1 = as1[0] + xlntaur * as1[1]
+    fs2 = as2[0] + xlntaur * as2[1]
+    del xlntaur
+    trdown = da.exp(-taur / mus)
+    trup = da.exp(-taur / muv)
+    xitm1 = (1.0 - trdown * trup) / 4.0 / (mus + muv)
+    xitm2 = (1.0 - trdown) * (1.0 - trup)
+    xitot1 = xph1 * (xitm1 + xitm2 * fs0)
+    xitot2 = xph2 * (xitm1 + xitm2 * fs1)
+    xitot3 = xph3 * (xitm1 + xitm2 * fs2)
+    rhoray = xitot1 * xcos1 + xitot2 * xcos2 * 2.0 + xitot3 * xcos3 * 2.0
+
+    if isinstance(height, xr.DataArray):
+        def _sphalb_index(sphalb0, index_arr):
+            # FIXME: if/when dask can support lazy index arrays then remove this
+            return sphalb0[index_arr]
+        sphalb_delayed = dask.delayed(_sphalb_index)(sphalb0, (taur / TAUSTEP4SPHALB + 0.5).astype(np.int32))
+        sphalb = da.from_delayed(sphalb_delayed, taur.shape, dtype=sphalb0.dtype)
```

Comment:
I thought about this while working on a separate feature. I think you could get this to work with `map_blocks` if you provide `index_arr` as the first argument, specify the `chunks=taur.chunks`, and most importantly before passing the `sphalb0` array do `sphalb0 = sphalb0.rechunk(sphalb0.shape)`. We can talk about this in person later.